### PR TITLE
Add support for ios_*_test_suite rules as top level targets

### LINKED ISSUE #1956: Bug: Test suite targets not supported as top level targets
### Description

If you hit [this code path](https://github.com/bazel-ios/rules_ios/blob/master/rules/test.bzl#L128) in `rules_ios` you'll get errors during project generation:
```sh
ERROR: /private/var/tmp/_bazel_thiago/122885c1fe4a2c6ed7635584956dfc9d/rules_xcodeproj/build_output_base/external/rules_xcodeproj_generated/generator/BUILD:6:10: in xcodeproj rule @rules_xcodeproj_generated//generator:generator: 
Traceback (most recent call last):
        File "/private/var/tmp/_bazel_thiago/122885c1fe4a2c6ed7635584956dfc9d/rules_xcodeproj/build_output_base/external/rules_xcodeproj/xcodeproj/internal/xcodeproj_rule.bzl", line 1390, column 26, in _xcodeproj_impl
                i = [_process_dep(dep) for dep in targets]
        File "/private/var/tmp/_bazel_thiago/122885c1fe4a2c6ed7635584956dfc9d/rules_xcodeproj/build_output_base/external/rules_xcodeproj/xcodeproj/internal/xcodeproj_rule.bzl", line 121, column 13, in _process_dep
                fail("""
Error in fail: 
'@//tests/macos/xcodeproj:Single-Application-RunnableTestSuite' is not a top-level target, but was listed in `top_level_targets`. Only list top-level targets (e.g. binaries, apps, tests, or distributable frameworks) in `top_level_targets`. Schemes and `focused_targets`/`unfocused_targets` can refer to dependencies of targets listed in `top_level_targets`, and don't need to be listed in `top_level_targets` themselves.

If you feel this is an error, and `test_suite` targets should be recognized as top-level targets, file a bug report here: https://github.com/MobileNativeFoundation/rules_xcodeproj/issues/new?template=bug.md
```

### Reproduction steps

* Checkout this branch: [thiago/repro-test-suite-not-supported-rules-xcodeproj](https://github.com/bazel-ios/rules_ios/compare/master...thiago/repro-test-suite-not-supported-rules-xcodeproj)
* Run `bazel run tests/ios/app:foo`
* You should hit the same err reported here

### Expected behavior

[ios_unit_test_suite](https://github.com/bazelbuild/rules_apple/blob/master/doc/rules-ios.md#ios_unit_test_suite) and [ios_ui_test_suite](https://github.com/bazelbuild/rules_apple/blob/master/doc/rules-ios.md#ios_ui_test_suite) can be passed as top level targets.

### rules_xcodeproj version

1.3.3

### Xcode version

14.2

### Bazel version

6.0.0

### rules_apple version

935d5ad80300578f35833db26f124f20aeda9cba

### rules_swift version

17e20f7edf27e647f1b45f11ed75d51c17820c3b

### Additional information

Happy to take this myself, any feedback and general guidance appreciated on how to approach adding support for it.

### PR DESCRIPTION
**Pull request by thiagohmcruz**
Work towards #1956 (will close #1968 in favour of the changes in this PR).

#2184 enables `test_suite` targets to be used as top level targets, the changes in this PR do the same for `ios_*_test_suite` targets.

* Track the `skip_type` during target generation.
* Use the value of `skip_type` to conditionally create the replacement labels.
    * Specifically, only do work if `skip_type` == `apple_test_bundle` and in that case removes the `rules_apple` internal suffix `.__internal__.__test_bundle` from the replacement label. This allows the custom scheme logic to match the provided `ios_*_test_suite` target label with the values in `replacement_labels` and do the correct replacement when creating the schemes in Xcode.

### CODE REVIEW

**Comment by thiagohmcruz** [comment]
@luispadron I'm adding tests but I'd appreciate if you could test this one and let me know how it goes 🙇

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 1]
```diff
@@ -88,20 +97,26 @@ def _should_skip_target(*, ctx, target):
         target: The `Target` to check.
 
     Returns:
-        `True` if `target` should be skipped for target generation.
+        A `tuple` with two elements:
+
+        -   A `bool` indicating if `target` should be skipped for target generation.
+        -   The `skip_type` for this target.
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:103]: No need for the tuple. Can just return the `skip_type` if it should be skipped, or `None` otherwise.

**Comment by luispadron** [comment]
@thiagohmcruz nice! im out until Tuesday but maybe @jszumski has time before then.

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 1 above)
Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:103]: Right, I guess if there is a skip type it should indeed be what implies a skip is needed, ok I'll update this

Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl]: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2196/commits/892ecded45235c59f6c316480d3ca733e3b1242e

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 2]
```diff
@@ -266,6 +278,20 @@ def _skip_target(
         if attr in deps_attrs and info.xcode_target
     ]
 
+    def _target_replacement_label(info):
+        if not info.xcode_target:
+            return target.label
+        if _APPLE_INTERNAL_TEST_BUNDLE_SUFFIX not in info.xcode_target.label.name:
```
Comment (thiagohmcruz) [inline xcodeproj/internal/xcodeprojinfo.bzl:284]: We also discussed log a warning if the suffix is not present, so I'll probably move the skip type check up and add the warning here

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 3]
```diff
@@ -97,27 +97,22 @@ def _get_skip_type(*, ctx, target):
...
-        targets.is_test_bundle(
-            target = target,
-            deps = getattr(ctx.rule.attr, "deps", None),
-        ),
-        skip_type.apple_test_bundle,
+    is_test_bundle = targets.is_test_bundle(
+        target = target,
+        deps = getattr(ctx.rule.attr, "deps", None),
     )
+    return skip_type.apple_test_bundle if is_test_bundle else None
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl]:
```suggestion
    if targets.is_test_bundle(
        target = target,
        deps = getattr(ctx.rule.attr, "deps", None),
    ):
        return skip_type.apple_test_bundle
    
    return None
```

**Review by brentleyjones — APPROVED** [review]

**Comment by luispadron** [comment]
Had a chance to test, I'm seeing the previous issue resolved (where running one of the targets in the test suite would run on all the devices defied in `runner`) it now correctly runs on just the Xcode selected device, even if it doesn't match the runner 👍🏼 .

Some things we should follow-up on:

- `test_suite` in a custom scheme breaks generation which is different than the other `top-level` target behavior. If it's not much more work I'd consider adding support for schemes in this PR/before next release so it's a smoother change.
- From a user point of view, I'd like the targets to be merged into a single test target if they only differ by `runner`, in our project these all test the same thing: 
  
  Screenshot 2023-05-16 at 12 00 17 PM

**Review by luispadron — APPROVED** [review]

**Comment by brentleyjones** [comment]
It shouldn't show as all those various versions anymore.

**Comment by thiagohmcruz** [comment]
@luispadron 

>test_suite in a custom scheme breaks generation which is different than the other top-level target behavior. If it's not much more work I'd consider adding support for schemes in this PR/before next release so it's a smoother change.

`test_suite` in custom schemes is currently not working indeed, there is a [follow up](test_suite in a custom scheme breaks generation which is different than the other top-level target behavior. If it's not much more work I'd consider adding support for schemes in this PR/before next release so it's a smoother change.) already that I'm planning to work on next (and yep we should have that before cutting a release)

>From a user point of view, I'd like the targets to be merged into a single test target if they only differ by runner, in our project these all test the same thing:

These should've been "merged", let me see if I can repro

**Comment by brentleyjones** [comment]
Ahh, then those won't be renamed.

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 3 above)
Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl]: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2196/commits/bced9a71e84979eb3d7af74eca1490921aca6b1d

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 2 above)
Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl]: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2196/commits/2044fa8c9e816ed30e0f0aed1637b2a1755079b9

**Comment by thiagohmcruz** [comment]
We might want to squash this 😅

**Comment by brentleyjones** [comment]
We always squash on merge 😄.

File: examples/integration/iOSApp/Test/SwiftUnitTests/BUILD [hunk 4]
```diff
@@ -1,4 +1,5 @@
-load("@build_bazel_rules_apple//apple:ios.bzl", "ios_unit_test")
+load("@build_bazel_rules_apple//apple:ios.bzl", "ios_unit_test", "ios_unit_test_suite")
+load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
```
Comment (brentleyjones) [inline examples/integration/iOSApp/Test/SwiftUnitTests/BUILD:2]:
```suggestion
load(
    "@build_bazel_rules_apple//apple:ios.bzl",
    "ios_unit_test",
    "ios_unit_test_suite",
)
load(
    "@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl",
    "ios_test_runner",
)
```

File: examples/integration/iOSApp/Test/UITests/BUILD [hunk 5]
```diff
@@ -1,5 +1,6 @@
-load("@build_bazel_rules_apple//apple:ios.bzl", "ios_ui_test")
+load("@build_bazel_rules_apple//apple:ios.bzl", "ios_ui_test", "ios_ui_test_suite")
 load("@build_bazel_rules_swift//swift:swift.bzl", "swift_library")
+load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
```
Comment (brentleyjones) [inline examples/integration/iOSApp/Test/UITests/BUILD:3]:
```suggestion
load(
    "@build_bazel_rules_apple//apple:ios.bzl",
    "ios_ui_test",
    "ios_ui_test_suite",
)
load(
    "@build_bazel_rules_swift//swift:swift.bzl",
    "swift_library",
)
load(
    "@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl",
    "ios_test_runner",
)
```

File: examples/integration/test/fixtures/bwb.xcodeproj/rules_xcodeproj/external.xcfilelist [hunk 6]
```diff
@@ -18,4 +18,7 @@ $(BAZEL_EXTERNAL)/com_google_google_maps/GoogleMapsCore.xcframework/ios-arm64_x8
 $(BAZEL_EXTERNAL)/examples_command_line_external/ExternalFramework.framework
 $(BAZEL_EXTERNAL)/examples_command_line_external/ImportableLibrary/Library.h
 $(BAZEL_EXTERNAL)/examples_command_line_external/ImportableLibrary/libImportableLibrary.a
+$(BAZEL_EXTERNAL)/examples_ios_app_external/Assets.xcassets
```
Comment (brentleyjones) [inline examples/integration/test/fixtures/bwb.xcodeproj/rules_xcodeproj/external.xcfilelist:21]: Interesting that this appears now.

File: examples/rules_ios/iOSApp/Test/ObjCUnitTests/BUILD [hunk 7]
```diff
@@ -2,6 +2,7 @@ load(
     "@build_bazel_rules_ios//rules:test.bzl",
     rules_ios_ios_unit_test = "ios_unit_test",
 )
+load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
```
Comment (brentleyjones) [inline examples/rules_ios/iOSApp/Test/ObjCUnitTests/BUILD:5]:
```suggestion
load(
    "@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl",
    "ios_test_runner",
)
```

File: examples/rules_ios/iOSApp/Test/SwiftUnitTests/BUILD [hunk 8]
```diff
@@ -1,4 +1,5 @@
 load("@build_bazel_rules_ios//rules:test.bzl", rules_ios_ios_unit_test = "ios_unit_test")
+load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
```
Comment (brentleyjones) [inline examples/rules_ios/iOSApp/Test/SwiftUnitTests/BUILD:2]:
```suggestion
load(
    "@build_bazel_rules_ios//rules:test.bzl",
    rules_ios_ios_unit_test = "ios_unit_test",
)
load(
    "@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl",
    "ios_test_runner",
)
```

File: examples/rules_ios/iOSApp/Test/UITests/BUILD [hunk 9]
```diff
@@ -1,4 +1,5 @@
 load("@build_bazel_rules_ios//rules:test.bzl", rules_ios_ios_ui_test = "ios_ui_test")
+load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
```
Comment (brentleyjones) [inline examples/rules_ios/iOSApp/Test/UITests/BUILD:2]:
```suggestion
load(
    "@build_bazel_rules_ios//rules:test.bzl",
    rules_ios_ios_ui_test = "ios_ui_test",
)
load(
    "@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl",
    "ios_test_runner",
)
```

File: examples/integration/test/fixtures/bwb.xcodeproj/rules_xcodeproj/external.xcfilelist (hunk 6 above)
Comment (thiagohmcruz) [reply examples/integration/test/fixtures/bwb.xcodeproj/rules_xcodeproj/external.xcfilelist:21]: Indeed, not idea if this is surfacing something we should investigate?

Comment (brentleyjones) [reply examples/integration/test/fixtures/bwb.xcodeproj/rules_xcodeproj/external.xcfilelist:21]: Now that I think about it, it's probably just you adding `"//iOSApp/Test/UITests:iOSAppUITests"` to the targets. Can you do that as a stand alone PR that merges before this, to keep the noise down?

Comment (thiagohmcruz) [reply examples/integration/test/fixtures/bwb.xcodeproj/rules_xcodeproj/external.xcfilelist:21]: Yep, definitely, will put up that PR in a few. Btw noticed that `//iOSApp/Test/UITests:iOSAppUITests` was also not set as top level target in the `rules_ios` example, should be put both in a separate PR then? Or one for each?

Comment (thiagohmcruz) [reply examples/integration/test/fixtures/bwb.xcodeproj/rules_xcodeproj/external.xcfilelist:21]: I'll put both in separate commits, PR coming soon

Comment (thiagohmcruz) [reply examples/integration/test/fixtures/bwb.xcodeproj/rules_xcodeproj/external.xcfilelist:21]: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2202

**Comment by luispadron** [comment]
> Ahh, then those won't be renamed.

Understand this change now more since it's based on `rules_apple` naming conventions, makes sense. We'll probably go that route unless ya'll have thoughts on how to support this better in Bazel:

- Running snapshot tests for all supported devices that are tested via a single label like: `bazel test //Foo:Snapshots` (runs 4 test targets with different runner)
- Running tests for one supported device via a single label like: `bazel test //Foo:SnapshotTests --config=iphone_13` (runs 1 test target with specific run, skips rest)
  - `runners` doesn't support `select` so we do this using the native `test_suite` rule along with `target_compatible_with`

We're probably going to give up allowing running on a single device and move to using the apple test suite rule instead. Folks can run via Xcode to target a device.

File: examples/integration/xcodeproj_targets.bzl [hunk 10]
```diff
@@ -153,4 +156,23 @@ def get_xcode_schemes():
...
+                },
+                targets = [
+                    "//iOSApp/Test/ObjCUnitTests:iOSAppObjCUnitTestSuite",
+                    "//iOSApp/Test/SwiftUnitTests:iOSAppSwiftUnitTestSuite",
+                ],
+                post_actions = [
+                    xcode_schemes.pre_post_action(
+                        name = "Run After Tests",
+                        script = "echo \"Hi\"",
+                        expand_variables_based_on = "//iOSApp/Test/SwiftUnitTests:iOSAppSwiftUnitTests",
```
Comment (thiagohmcruz) [inline examples/integration/xcodeproj_targets.bzl:173]: Oops, this should probably be one of the `targets` passed above and not this. 

ps: a custom scheme for raw `test_suite` will be added as well but in a follow up (some remaining things to implement there in order to support it)

Comment (thiagohmcruz) [reply examples/integration/xcodeproj_targets.bzl]: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2196/commits/77f8c029649d0f4ebe037dc443d6595c40798df7

**Comment by thiagohmcruz** [comment]
Will rebase and make tests `manual` in this PR once #2205 lands.

**Comment by thiagohmcruz** [comment]
> * target_compatible_with

@luispadron The way I see it the set of PRs I'm landing mostly address two use cases:

* Raw `test_suite` => maps all targets 1-1 to Xcode
* `rules_apple`'s `ios_*_test_suite` => "merge" into a single target scheme allowing Xcode to be the runner (this is possible due to known naming pattern and impl details in `rules_apple` specifically)

We should def discuss if your use case is something we want to support but I'd say it's out of scope for this PR. Specifically:

>Running snapshot tests for all supported devices that are tested via a single label like: bazel test //Foo:Snapshots (runs 4 test targets with different runner)

* `test_suite` with any tests + "some configuration upfront?" => run all tests from a single label (should be achievable [via alias_labels](https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2184#issuecomment-1544025699) which would allow `test_suite` to be set as a custom scheme? Not sure yet, need to think more about this)

>Running tests for one supported device via a single label like: bazel test //Foo:SnapshotTests --config=iphone_13 (runs 1 test target with specific run, skips rest)

I think this would be a special case of the item above, same thing but ignoring all runners when `test_suite` is a custom scheme and instead of running all underlying tests "merge" them so Xcode can be the runner for you to select say `iPhone 13` 

Just thinking out loud, I don't have a good answer to this atm. Will iterate on `alias_labels` next and see if the above applies or not.

**Pull request merged**